### PR TITLE
feat: print mutation result in json

### PR DESCRIPTION
**Pull request by fregataa**
`--output=json` option offers jsonified result of mutation.

add `print_mutation_result` abstract method in `client.output.types.BaseOutputHandler`

image

in the future, [Backend.ai Test](https://github.com/lablup/backend.ai-test) should be updated.

### CODE REVIEW

File: src/ai/backend/client/cli/admin/domain.py [hunk 1]
```diff
@@ -90,11 +91,17 @@ def add(name, description, inactive, total_resource_slots,
         if not data['ok']:
             print_fail('Domain creation has failed: {0}'.format(data['msg']))
             sys.exit(1)
-        item = data['domain']
-        print('Domain name {0} is created.'.format(item['name']))
+        ctx.output.print_mutation_result(
+            data,
+            item_name='domain',
+            add_info={
+                'method': sys._getframe().f_code.co_name,
```
Comment (achimnol) [inline src/ai/backend/client/cli/admin/domain.py:98]: This is too much hacky, because it uses a CPython-specific private API. (CPython is just one of many Python language implementations, though it is the reference implementation.)

To extract the command name, we could use Click's API instead.

```python
@click.pass_context
...
def add(click_ctx: click.Context, ...):
    ...
    print(click_ctx.command.name)
```

There are many more rich information about the current command in [the Click's context object](https://click.palletsprojects.com/en/8.0.x/api/#context).

It is better to use publicly available APIs whenever possible.

Comment (achimnol) [reply src/ai/backend/client/cli/admin/domain.py:98]: BTW, since the purpose of JSON output is to mechanically automate user-side scripting and the caller already knows what command it called because it called, I don't think this is a strictly required information to include in every mutation's JSON output.

Please justify the scenarios where this field is useful or remove it.

File: src/ai/backend/client/output/json.py [hunk 2]
```diff
@@ -122,6 +123,29 @@ def print_paginated_list(
             **_json_opts,
         ))
 
+    def print_mutation_result(
+        self,
+        item: Mapping[str, Any],
+        item_name: Optional[str] = None,
+        add_info: Mapping = {},
```
Comment (achimnol) [inline src/ai/backend/client/output/json.py:130]: Rename `add_info` to `extra_info` or `additional_info` to make "add" to stand out as an adjective instead of a verb.

**Review by achimnol — CHANGES_REQUESTED** [review]

File: src/ai/backend/client/cli/admin/domain.py (hunk 1 above)
Comment (fregataa) [reply src/ai/backend/client/cli/admin/domain.py:98]: I removed those function name field.
I simply wrote it because CLI shows what function is called with --output=console option.

**Review by achimnol — CHANGES_REQUESTED** [review]
Now it looks good for the successful execution results.

Though, to fully automate the user-side scripting, we also need to wrap `print_fail()` or `print_error()` calls in the Click command functions by extending the output framework like your addition of `print_mutation_result()`.

For instance,
```python
    with Session() as session:
        try:
            data = session.KeyPair.deactivate(access_key)
        except Exception as e:
            print_error(e)
            sys.exit(1)
        if not data['ok']:
            print_fail('KeyPair deactivation has failed: {0}'.format(data['msg']))
            sys.exit(1)
        ctx.output.print_mutation_result(
            data,
            extra_info={
                'access_key': access_key,
            },
        )
```
could be rewritten to
```python
    with Session() as session:
        try:
            data = session.KeyPair.deactivate(access_key)
        except Exception as e:
            # add a new output handler for errors
            ctx.output.print_mutation_error(
                e,
                object_name="keypair",
                action_name="deactivation",
            )
            sys.exit(1)
        # handle the case when not data['ok'] in print_mutation_result()
        ctx.output.print_mutation_result(
            data,
            object_name="keypair",
            action_name="deactivation",
            extra_info={
                'access_key': access_key,
            },
        )
```

Note: explicitly specifying `object_name` and `action_name` is just an idea for console outputs to print "what the user has tried to do", which is not required for json outputs.

File: src/ai/backend/client/output/json.py [hunk 3]
```diff
@@ -122,6 +123,52 @@ def print_paginated_list(
...
         ))
 
+    def print_mutation_result(
+        self,
+        item: Mapping[str, Any],
+        item_name: Optional[str] = None,
+        action_name: Optional[str] = None,
+        extra_info: Mapping = {},
+    ) -> None:
+        t = {
```
Comment (achimnol) [inline src/ai/backend/client/output/json.py:133]: Let's use a more sensible variable name like `data` in both `print_mutation_result()` and `print_mutation_error()`.

File: src/ai/backend/client/cli/admin/scaling_group.py [hunk 4]
```diff
@@ -142,17 +153,31 @@ def update(name, description, inactive,
...
+            ctx.output.print_mutation_error(
+                e,
+                item_name='scaling_group',
+                action_name='update',
+            )
             sys.exit(1)
         if not data['ok']:
-            print_fail('Scaling group update has failed: {0}'.format(data['msg']))
+            ctx.output.print_mutation_error(
+                msg=data['msg'],
```
Comment (achimnol) [inline src/ai/backend/client/cli/admin/scaling_group.py:164]: Currently, the manager-side error message generation routine just quotes the mutation class name, resulting in messages like "no matching modifyscalinggroup".

Could you make an idea to improve it *in the manager side*? (e.g., "no matching scaling group")
Let's make a separate PR to the manager for this.

**Review by achimnol — CHANGES_REQUESTED** [review]
Overall, looks great.
Added a request for a minor fix and a suggestino for manager-side improvement.

File: src/ai/backend/client/cli/admin/scaling_group.py (hunk 4 above)
Comment (fregataa) [reply src/ai/backend/client/cli/admin/scaling_group.py:164]: it would be great to inform the reason of the error in the manager side.
this is an another issue from this pr?

**Review by achimnol — APPROVED** [review]
Let's handle prettifying object names in error messages as a separate PR.
